File: backend/src/schemas/production_progress.py
from typing import Literal

from schemas.production_base import ProductionModel
from schemas.production_work_orders import WorkOrderResponse


class AssemblyMaterialArrivalResponse(ProductionModel):
    material_type: Literal["part", "assembly"]
    material_no: str
    material_name: str
    task_quantity: int
    arrived_quantity: int


class ProductionTaskWorkOrderContributionResponse(ProductionModel):
    work_order_id: int
    quantity: int


class ProductionTaskBatchContributionResponse(ProductionModel):
    batch_id: int
    quantity: int


class ProductionTaskProcessingStatusResponse(ProductionModel):
    status: Literal[
        "not_started",
        "processing",
        "submitted_qc",
        "rework",
        "completed",
        "exception",
        "department_completed",
    ]
    label: str
    action: Literal[
        "create_work_order",
        "view_work_orders",
        "none",
    ]
    quantity: int
    procedure_names: list[str]
    repository_ids: list[int]
    related_work_orders: list[ProductionTaskWorkOrderContributionResponse]
    related_batches: list[ProductionTaskBatchContributionResponse]
    creation_mode: Literal["repository", "assembly_initial"] | None


class DepartmentProductionProgressResponse(ProductionModel):
    production_plan_item_id: int
    customer_order_item_id: int
    production_item_id: int | None
    flow_node_id: str
    plan_status: Literal["draft", "confirmed", "cancelled", "completed"]
    part_no: str
    part_name: str
    processing_workshop_id: int
    processing_workshop: str
    task_quantity: int
    arrived_quantity: int
    material_arrivals: list[AssemblyMaterialArrivalResponse]
    processing_statuses: list[ProductionTaskProcessingStatusResponse]
    completed_quantity: int
    remark: str


class DepartmentProductionProgressEnvelope(ProductionModel):
    data: list[DepartmentProductionProgressResponse]
    total: int


class ProductionProgressWorkOrderResponse(ProductionModel):
    id: int
    work_order_no: str
    work_order_type: Literal["standard", "assembly", "supplier_processing"]
    is_temporary: bool
    qc_available: bool
    direct_result_allowed: bool
    output_unit_quantity: int
    procedure_name: str
    workshop_name: str
    worker_name: str | None
    work_order_quantity: int
    quantity: int
    processed_quantity: int
    submitted_quantity: int
    pending_qc_quantity: int
    completed_quantity: int
    rework_quantity: int
    scrap_quantity: int
    lost_quantity: int
    status: Literal["open", "closed", "cancelled"]
    created_at: str
    closed_at: str | None
    work_order: WorkOrderResponse


class ProductionProgressProcedureCardResponse(ProductionModel):
    card_key: str
    card_type: Literal["process", "assembly"]
    sort_order: int
    flow_node_id: str
    procedure_id: int | None
    card_name: str
    department_code: str
    department_name: str
    workshop_id: int | None
    workshop_name: str
    procedure_name: str
    status: Literal["not_arrived", "ready", "processing", "pending_qc", "completed", "exception"]
    task_quantity: int
    arrived_quantity: int
    processing_quantity: int
    ready_for_qc_quantity: int
    pending_qc_quantity: int
    completed_quantity: int
    rework_quantity: int
    scrap_quantity: int
    lost_quantity: int
    work_orders: list[ProductionProgressWorkOrderResponse]


class ProductionProgressItemDetailResponse(ProductionModel):
    production_plan_item_id: int
    production_item_ids: list[int]
    customer_order_no: str
    factory_code: str
    product_name: str
    part_no: str
    part_name: str
    plan_status: Literal["draft", "confirmed", "cancelled", "completed"]
    task_quantity: int
    cards: list[ProductionProgressProcedureCardResponse]
    work_orders: list[ProductionProgressWorkOrderResponse]

File: frontend/src/features/production/domain/productionProgress.ts
import type { ProductionPlanStatus } from '@/features/customer-orders'
import type { WorkOrder, WorkOrderStatus } from './types'

export type ProductionProgressCardStatus =
  | 'not_arrived'
  | 'ready'
  | 'processing'
  | 'pending_qc'
  | 'completed'
  | 'exception'

export type DepartmentProductionProgressItem = {
  production_plan_item_id: number
  customer_order_item_id: number
  customer_order_no: string
  production_item_id: number | null
  flow_node_id: string
  plan_status: ProductionPlanStatus
  part_no: string
  part_name: string
  processing_workshop_id: number
  processing_workshop: string
  task_quantity: number
  arrived_quantity: number
  material_arrivals: AssemblyMaterialArrival[]
  processing_statuses: ProductionTaskProcessingStatus[]
  completed_quantity: number
  remark: string
}

export type ProductionTaskProcessingStatusCode =
  | 'not_started'
  | 'processing'
  | 'submitted_qc'
  | 'rework'
  | 'completed'
  | 'exception'
  | 'department_completed'

export type ProductionTaskProcessingStatus = {
  status: ProductionTaskProcessingStatusCode
  label: string
  action: 'create_work_order' | 'view_work_orders' | 'none'
  quantity: number
  procedure_names: string[]
  repository_ids: number[]
  related_work_orders: Array<{ work_order_id: number; quantity: number }>
  related_batches: Array<{ batch_id: number; quantity: number }>
  creation_mode: 'repository' | 'assembly_initial' | null
}

export type AssemblyMaterialArrival = {
  material_type: 'part' | 'assembly'
  material_no: string
  material_name: string
  task_quantity: number
  arrived_quantity: number
}

export type ProductionProgressWorkOrder = {
  id: number
  work_order_no: string
  work_order_type: 'standard' | 'assembly' | 'supplier_processing'
  is_temporary: boolean
  qc_available: boolean
  direct_result_allowed: boolean
  output_unit_quantity: number
  procedure_name: string
  workshop_name: string
  worker_name: string | null
  work_order_quantity: number
  quantity: number
  processed_quantity: number
  submitted_quantity: number
  pending_qc_quantity: number
  completed_quantity: number
  rework_quantity: number
  scrap_quantity: number
  lost_quantity: number
  status: WorkOrderStatus
  created_at: string
  closed_at: string | null
  work_order: WorkOrder
}

export type ProductionProgressProcedureCard = {
  card_key: string
  card_type: 'process' | 'assembly'
  sort_order: number
  flow_node_id: string
  procedure_id: number | null
  card_name: string
  department_code: string
  department_name: string
  workshop_id: number | null
  workshop_name: string
  procedure_name: string
  status: ProductionProgressCardStatus
  task_quantity: number
  arrived_quantity: number
  processing_quantity: number
  ready_for_qc_quantity: number
  pending_qc_quantity: number
  completed_quantity: number
  rework_quantity: number
  scrap_quantity: number
  lost_quantity: number
  work_orders: ProductionProgressWorkOrder[]
}

export type ProductionProgressItemDetail = {
  production_plan_item_id: number
  production_item_ids: number[]
  customer_order_no: string
  factory_code: string
  product_name: string
  part_no: string
  part_name: string
  plan_status: ProductionPlanStatus
  task_quantity: number
  cards: ProductionProgressProcedureCard[]
  work_orders: ProductionProgressWorkOrder[]
}
